pkg/instancetype/apply/cpu.go

```diff
@@ -0,0 +1,157 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+package apply
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	v1beta1 "kubevirt.io/api/instancetype/v1beta1"
+
+	preferenceApply "kubevirt.io/kubevirt/pkg/instancetype/preference/apply"
+)
+
+func applyCPU(
+	field *k8sfield.Path,
+	instancetypeSpec *v1beta1.VirtualMachineInstancetypeSpec,
+	preferenceSpec *v1beta1.VirtualMachinePreferenceSpec,
+	vmiSpec *virtv1.VirtualMachineInstanceSpec,
+) Conflicts {
+	if vmiSpec.Domain.CPU == nil {
+		vmiSpec.Domain.CPU = &virtv1.CPU{}
+	}
+
+	// If we have any conflicts return as there's no need to apply the topology below
+	if conflicts := validateCPU(field, instancetypeSpec, vmiSpec); len(conflicts) > 0 {
+		return conflicts
+	}
+
+	if vmiSpec.Domain.CPU.Model == "" && instancetypeSpec.CPU.Model != nil {
+		vmiSpec.Domain.CPU.Model = *instancetypeSpec.CPU.Model
+	}
+
+	if instancetypeSpec.CPU.DedicatedCPUPlacement != nil {
+		vmiSpec.Domain.CPU.DedicatedCPUPlacement = *instancetypeSpec.CPU.DedicatedCPUPlacement
+	}
+
+	if instancetypeSpec.CPU.IsolateEmulatorThread != nil {
+		vmiSpec.Domain.CPU.IsolateEmulatorThread = *instancetypeSpec.CPU.IsolateEmulatorThread
+	}
+
+	if vmiSpec.Domain.CPU.NUMA == nil && instancetypeSpec.CPU.NUMA != nil {
+		vmiSpec.Domain.CPU.NUMA = instancetypeSpec.CPU.NUMA.DeepCopy()
+	}
+
+	if vmiSpec.Domain.CPU.Realtime == nil && instancetypeSpec.CPU.Realtime != nil {
+		vmiSpec.Domain.CPU.Realtime = instancetypeSpec.CPU.Realtime.DeepCopy()
+	}
+
+	if instancetypeSpec.CPU.MaxSockets != nil {
+		vmiSpec.Domain.CPU.MaxSockets = *instancetypeSpec.CPU.MaxSockets
+	}
+
+	applyGuestCPUTopology(instancetypeSpec.CPU.Guest, preferenceSpec, vmiSpec)
+
+	return nil
+}
+
+func applyGuestCPUTopology(vCPUs uint32, preferenceSpec *v1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
+	// Apply the default topology here to avoid duplication below
+	vmiSpec.Domain.CPU.Cores = 1
+	vmiSpec.Domain.CPU.Sockets = 1
+	vmiSpec.Domain.CPU.Threads = 1
+
+	if vCPUs == 1 {
+		return
+	}
+
+	switch preferenceApply.GetPreferredTopology(preferenceSpec) {
+	case v1beta1.DeprecatedPreferCores, v1beta1.Cores:
+		vmiSpec.Domain.CPU.Cores = vCPUs
+	case v1beta1.DeprecatedPreferSockets, v1beta1.DeprecatedPreferAny, v1beta1.Sockets, v1beta1.Any:
+		vmiSpec.Domain.CPU.Sockets = vCPUs
+	case v1beta1.DeprecatedPreferThreads, v1beta1.Threads:
+		vmiSpec.Domain.CPU.Threads = vCPUs
+	case v1beta1.DeprecatedPreferSpread, v1beta1.Spread:
+		ratio, across := preferenceApply.GetSpreadOptions(preferenceSpec)
```

Comment:
Where do we validate again, that the requested combination of topology and the amount vCPUs is compatible with each other?

Reply:
https://github.com/kubevirt/kubevirt/blob/a83db2fb18d646cb37901ffe3f448864a2c991e2/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L267-L301 - That needs to be extracted into this package once this PR lands tbh